File: test/estimate/hessdiag/hessizero_saveall.jl
# Compute Hessian of posterior function evaluated at x (vector)
# if noisy, display error messages, results, etc.
# 11/12/01 translated by Marco DelNegro in matlab from Frank Schorfheide's program in gauss
function hessizero_saveall!{T<:FloatingPoint}(x::Vector{T}, model::AbstractModel, YY::Matrix{T}; noisy::Bool = false)

    update!(model.Θ, x)
    
    ## index of free parameters
    para_free = [!α.fixed for α in model.Θ]
    fpara_free = find(para_free)
    nfree = length(fpara_free)

    npara = length(x)
    ndx = 6
    dx =  exp(-(6:2:(6+(ndx-1)*2))')
    hessdiag = zeros(npara, npara, ndx)



    # Compute Diagonal elements first
    for seli = fpara_free'
        if noisy
            println("\nHessian element: ($seli, $seli)")
        end

        for k = 1:ndx
            paradx = copy(x)
            parady = copy(x)
            paradx[seli] = paradx[seli] + dx[k]
            parady[seli] = parady[seli] - dx[k]

            fx = posterior!(x, model, YY)
            fdx = posterior!(paradx, model, YY)
            fdy = posterior!(parady, model, YY)
            hessdiag[seli, seli, k] = -(2fx - fdx - fdy) / (dx[k])^2
        end
        
        if noisy
            values = reshape(hessdiag[seli, seli, :], 6, 1)
            println("Values: $values")
        end
    end

    # Now compute off-diagonal elements
    for i = 1:(nfree-1)
        seli = fpara_free[i]
        for j = (i+1):nfree
            selj = fpara_free[j]
            
            if noisy
                println("\nHessian element: ($seli, $selj)")
            end
            
            for k = 1:ndx
                paradx = copy(x)
                parady = copy(x)
                paradx[seli] = paradx[seli] + dx[k]
                parady[selj] = parady[selj] - dx[k]
                paradxdy = copy(paradx)
                paradxdy[selj] = paradxdy[selj] - dx[k]
                
                fx = posterior!(x, model, YY)
                fdx = posterior!(paradx, model, YY)
                fdy = posterior!(parady, model, YY)
                fdxdy = posterior!(paradxdy, model, YY)
                hessdiag[seli, selj, k] = -(fx - fdx - fdy + fdxdy) / (dx[k]*dx[k])
                hessdiag[selj, seli, k] = hessdiag[seli, selj, k]
            end
            
            if noisy
                values = reshape(hessdiag[seli, selj, :], 6, 1)
                println("Values: $values")
            end
        end
    end

    return hessdiag
end

File: test/estimate/hessdiag/hessizero_saveall.m
function hessdiag = hessizero_saveall(fcn,x,Verbose,varargin)
% computes hessian of function fcn (string) evaluated at x (vector)
% varargin are the other inputs of fcn
% if Verbose, display error messages , results, etc.
% 11/12/01 translated by Marco DelNegro in matlab from Frank Schorfheide's program in gauss

%% index of free parameters
para_free = 1-x(:,2);
fpara_free = find(para_free);
nfree = length(fpara_free);

%% actual max
x = x(:,1);

npara = length(x);
ndx = 6;
dx =  exp(-(6:2:(6+(ndx-1)*2))');
hessdiag = zeros(npara, npara, ndx);



% Compute Diagonal elements first
for seli = fpara_free'

	if Verbose; fprintf(1,'\n\n Hessian Element: (%2.2g %2.2g)',[seli,seli]); end;
	for i=1:ndx;
		paradx = x;
		parady = x;
		paradx(seli) = paradx(seli) + dx(i);
		parady(seli) = parady(seli) - dx(i);
        
        fx  = eval([fcn '(x,varargin{:})']);
		fdx = eval([fcn '(paradx,varargin{:})']);
		fdy = eval([fcn '(parady,varargin{:})']);
		hessdiag(seli, seli, i) = -( 2*fx - fdx - fdy)/(dx(i))^2; 
    end
  	if Verbose == 2; fprintf(1,'\n Values: %2.6f',hessdiag(seli, seli, :));     end;
end

% Now compute off-diagonal elements
for II = 1:(nfree-1);
   seli = fpara_free(II);	
   for JJ = II+1:nfree;
	selj = fpara_free(JJ);
    	if Verbose; fprintf(1,'\n\n Hessian Element: (%2.2g %2.2g)',[seli,selj]); end;
	    for i=1:ndx;
			paradx = x;
			parady = x;
			paradx(seli) = paradx(seli) + dx(i);
			parady(selj) = parady(selj) - dx(i);
			paradxdy = paradx;
			paradxdy(selj) = paradxdy(selj) - dx(i);
    	
            fx  = eval([fcn '(x,varargin{:})']);
    		fdx = eval([fcn '(paradx,varargin{:})']);
    		fdy = eval([fcn '(parady,varargin{:})']);
            fdxdy = eval([fcn '(paradxdy,varargin{:})']);
			hessdiag(seli, selj, i) = -( fx - fdx - fdy + fdxdy )/(dx(i)*dx(i));
            hessdiag(selj, seli, i) = hessdiag(seli, selj, i);
        end
    	if Verbose == 2; fprintf(1,'\n Values: %2.6f',hessdiag(seli, selj, :)); end;
    end
end

